File: memory_model/sources/mem_model.1.c
/*
* @@name:	mem_model.1c
* @@type:	C
* @@compilable:	yes
* @@linkable:	yes
* @@expect:	rt-error
* @@version:	omp_3.1
*/
#include <stdio.h>
#include <omp.h>

int main(){
  int x;

  x = 2;
  #pragma omp parallel num_threads(2) shared(x)
  {

    if (omp_get_thread_num() == 0) {
       #pragma omp atomic write
       x = 5;
    } else {
      int xval;
      #pragma omp atomic read
      xval = x;
    /* Print 1: xval can be 2 or 5 */
      printf("1: Thread# %d: x = %d\n", omp_get_thread_num(), xval);
    }

    #pragma omp barrier

    if (omp_get_thread_num() == 0) {
    /* Print 2 */
      printf("2: Thread# %d: x = %d\n", omp_get_thread_num(), x);
    } else {
    /* Print 3 */
      printf("3: Thread# %d: x = %d\n", omp_get_thread_num(), x);
    }
  }
  return 0;
}

File: memory_model/sources/mem_model.1.f90
! @@name:	mem_model.1
! @@type:	F-free
! @@compilable:	yes
! @@linkable:	yes
! @@expect:	rt-error
! @@version:	omp_3.1
PROGRAM MEMMODEL
  INCLUDE "omp_lib.h"      ! or USE OMP_LIB
  INTEGER X, XVAL

  X = 2
!$OMP PARALLEL NUM_THREADS(2) SHARED(X)

    IF (OMP_GET_THREAD_NUM() .EQ. 0) THEN
    !$OMP ATOMIC WRITE
       X = 5
    ELSE
    !$OMP ATOMIC READ
      XVAL = X
    ! PRINT 1: XVAL can be 2 or 5
      PRINT *,"1: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", XVAL
    ENDIF

 !$OMP BARRIER

    IF (OMP_GET_THREAD_NUM() .EQ. 0) THEN
    ! PRINT 2
      PRINT *,"2: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", X
    ELSE
    ! PRINT 3
      PRINT *,"3: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", X
    ENDIF

!$OMP END PARALLEL

END PROGRAM MEMMODEL
